File: py/3Sum.py
class Solution:
    # @return a list of lists of length 3, [[val1,val2,val3]]
    def threeSum(self, num):
        num.sort()
        n = len(num)
       
        lret = []
       
        for i in range(n):
            
            if num[i] > 0:
                break
            
            if i != 0 and num[i] == num[i-1]:
                continue
            
            l = i+1
            h = n-1
            target = num[i]
            
            while l < h:
                if num[l] + num[h] + target > 0:
                    h -= 1
                    
                elif num[l] + num[h] + target < 0:
                    l += 1
                    
                elif num[l] + num[h] + target == 0:
                    lret.append([target,num[l],num[h]])
                    l += 1
                    h -= 1
                    while l < h and num[l] == num[l-1]:
                        l += 1
                    while l < h and num[h] == num[h+1]:
                        h -= 1
        return lret

File: cpp/sum_problem/3sum.cpp
class Solution {
public:
    vector<vector<int> > threeSum(vector<int> &num) {
        
        vector<vector<int>> vret;
        
        if(num.size() == 0)
            return vret;
            
        sort(num.begin(),num.end());
        
        for(int i = 0;i < num.size()-1;i++)
        {
            //remove duplecated results
            if(i > 0 && num[i] == num[i-1]) 
                continue;    
            
            
            int p = i+1;
            int j = num.size() - 1;
        
            while(p < j)
            {
                if(num[p] + num[j] + num[i] == 0)
                {
                    vector<int> tmp;
                    tmp.push_back(num[i]);
                    tmp.push_back(num[p]);
                    tmp.push_back(num[j]);
                    
                    vret.push_back(tmp);
                    p += 1;
                    j -= 1;
                    
                    //remove duplecated results
                    while(p < num.size() && num[p]==num[p - 1]) 
                        p++;    
                    while(j >= 0 && num[j] == num[j + 1]) 
                        j--;
                }
                else if(num[p] + num[j] + num[i] > 0)
                    j -= 1;
                else
                    p += 1;
            }
        }
        
        return vret;
    }
};